File: typescript/src/security.ts
/**
 * Security utilities for AgentSentinel.
 *
 * @module security
 */

/** Fine-grained security configuration used by {@link AgentPolicy}. */
export interface SecurityConfigOptions {
  /**
   * Regular-expression patterns whose matches are replaced with `[REDACTED]`
   * before writing to audit logs.  Defaults to common credential patterns
   * (API keys, passwords, bearer tokens, secrets).
   */
  redactPatterns?: string[];

  /**
   * Tools that are **always** blocked — they will never execute regardless of
   * approval decisions.  Supports {@link https://www.npmjs.com/package/micromatch micromatch}-style
   * glob wildcards (e.g. `"delete_*"`).
   */
  blockedTools?: string[];

  /**
   * Tools that always require explicit human approval even when not listed in
   * {@link AgentPolicyOptions.requireApproval}.  Defaults to a curated list
   * of high-risk operations (shell execution, file deletion, payments, etc.).
   */
  sensitiveTools?: string[];

  /**
   * Maximum number of characters captured per parameter in audit logs.
   * Prevents large payloads (file contents, model outputs) from bloating the
   * audit trail.  Default: `1000`.
   */
  maxParamLogSize?: number;

  /**
   * When `false` (the default) parameter values are truncated to
   * {@link maxParamLogSize} characters and sensitive patterns are redacted
   * before logging.  Set to `true` only in secure, controlled environments.
   */
  logFullParams?: boolean;
}

/** Immutable security configuration value object. */
export class SecurityConfig {
  readonly redactPatterns: readonly string[];
  readonly blockedTools: readonly string[];
  readonly sensitiveTools: readonly string[];
  readonly maxParamLogSize: number;
  readonly logFullParams: boolean;

  constructor(options: SecurityConfigOptions = {}) {
    this.redactPatterns = Object.freeze(
      options.redactPatterns ?? [
        'api[_-]?key["\']?\\s*[:=]\\s*["\']?[\\w-]+',
        'password["\']?\\s*[:=]\\s*["\']?[^\\s"\']+',
        'secret["\']?\\s*[:=]\\s*["\']?[\\w-]+',
        'token["\']?\\s*[:=]\\s*["\']?[\\w-]+',
        'bearer\\s+[\\w-]+',
      ]
    );
    this.blockedTools = Object.freeze(options.blockedTools ?? []);
    this.sensitiveTools = Object.freeze(
      options.sensitiveTools ?? [
        "execute_shell",
        "run_command",
        "delete_file",
        "rm_rf",
        "drop_table",
        "send_email",
        "post_tweet",
        "make_payment",
      ]
    );
    this.maxParamLogSize = options.maxParamLogSize ?? 1000;
    this.logFullParams = options.logFullParams ?? false;
  }
}

/**
 * Redact sensitive information from *text* using *patterns*.
 *
 * Each pattern is matched case-insensitively; matches are replaced with the
 * literal string `[REDACTED]`.
 */
export function redactSensitive(text: string, patterns: readonly string[]): string {
  let result = text;
  for (const pattern of patterns) {
    result = result.replace(new RegExp(pattern, "gi"), "[REDACTED]");
  }
  return result;
}

/**
 * Return `true` if *toolName* matches any glob pattern in *blockedList*.
 *
 * Uses a simple `*` wildcard check compatible with fnmatch semantics.
 */
export function isToolBlocked(toolName: string, blockedList: readonly string[]): boolean {
  return blockedList.some((pattern) => matchesGlob(toolName, pattern));
}

/**
 * Minimal glob matching supporting `*` (any characters) and `?` (one character).
 * Converts the pattern to a RegExp for matching.
 */
function matchesGlob(name: string, pattern: string): boolean {
  if (pattern === name) return true;
  // Escape regex special chars except * and ?
  const regexStr = pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replace(/\*/g, ".*")
    .replace(/\?/g, ".");
  return new RegExp(`^${regexStr}$`).test(name);
}

File: python/agentsentinel/security.py
"""Security utilities for AgentSentinel."""

from __future__ import annotations

import fnmatch
import re
from dataclasses import dataclass, field
from typing import List


@dataclass
class SecurityConfig:
    """Security configuration for AgentSentinel.

    Parameters
    ----------
    redact_patterns:
        Regular-expression patterns to redact from audit log output.
        Defaults to common credential patterns (API keys, passwords, tokens).
    blocked_tools:
        Tools that are **always** blocked — they will never execute,
        regardless of approvals.  Supports ``fnmatch``-style wildcards.
        Use this as an emergency kill-list for catastrophic operations.
    sensitive_tools:
        Tools that always require explicit human approval even if they are
        not listed in :attr:`.AgentPolicy.require_approval`.  Defaults to
        a curated list of high-risk operations (shell execution, file
        deletion, financial actions, etc.).
    max_param_log_size:
        Maximum number of characters captured per parameter in audit logs.
        Prevents large payloads (file contents, model outputs) from
        bloating the audit trail.
    log_full_params:
        When ``False`` (the default) parameter values are truncated to
        *max_param_log_size* characters and sensitive patterns are
        redacted before logging.  Set to ``True`` only in secure,
        controlled environments.
    """

    # Patterns to redact from audit logs (e.g., API keys, passwords)
    redact_patterns: List[str] = field(default_factory=lambda: [
        r'api[_-]?key["\']?\s*[:=]\s*["\']?[\w-]+',
        r'password["\']?\s*[:=]\s*["\']?[^\s"\']+',
        r'secret["\']?\s*[:=]\s*["\']?[\w-]+',
        r'token["\']?\s*[:=]\s*["\']?[\w-]+',
        r'bearer\s+[\w-]+',
    ])

    # Tools that are ALWAYS blocked (emergency kill list)
    blocked_tools: List[str] = field(default_factory=list)

    # Tools that require approval even if not in policy.require_approval
    sensitive_tools: List[str] = field(default_factory=lambda: [
        "execute_shell",
        "run_command",
        "delete_file",
        "rm_rf",
        "drop_table",
        "send_email",
        "post_tweet",
        "make_payment",
    ])

    # Maximum parameter size to log (prevents memory issues with large payloads)
    max_param_log_size: int = 1000

    # Whether to log full parameters (False = redact/truncate by default)
    log_full_params: bool = False


def redact_sensitive(text: str, patterns: List[str]) -> str:
    """Redact sensitive information from *text* using *patterns*.

    Each pattern is compiled with :data:`re.IGNORECASE` so credential
    names are matched regardless of capitalisation.

    Parameters
    ----------
    text:
        The string to sanitise.
    patterns:
        List of regular-expression patterns whose matches are replaced
        with the literal string ``'[REDACTED]'``.

    Returns
    -------
    str
        A copy of *text* with all sensitive matches replaced.
    """
    result = text
    for pattern in patterns:
        result = re.sub(pattern, "[REDACTED]", result, flags=re.IGNORECASE)
    return result


def is_tool_blocked(tool_name: str, blocked_list: List[str]) -> bool:
    """Return ``True`` if *tool_name* matches any entry in *blocked_list*.

    Matching uses :func:`fnmatch.fnmatch` so entries like ``"delete_*"``
    block any tool whose name starts with ``delete_``.

    Parameters
    ----------
    tool_name:
        The name of the tool being checked.
    blocked_list:
        Patterns to match against (exact names or fnmatch wildcards).
    """
    return any(fnmatch.fnmatch(tool_name, pattern) for pattern in blocked_list)
